src/WRings.sol

```diff
@@ -0,0 +1,268 @@
+    }
+
+    /// @notice max amount of tokens that could be withdrawn by a user
+    /// @param owner address for the owner of the shares being redeemed
+    /// @return maxAssets maximum number of assets owner is able to withdraw
+    function maxWithdraw(address owner) external view returns(uint256) {
+        return _convertToAssets(_balanceOf[owner]);
+    }
+
+    /// @notice max amout of shares a user could redeem
+    /// @param owner address for the owner of the shares being redeemed
+    /// @return maxShares 
+    function maxRedeem(address owner) external view returns(uint256) {
+        return _balanceOf[owner];
+    }
+
+    /// @notice used to simulate a deposit at the current block
+    /// @param assets amount of assets being deposited
+    /// @return shares amount of shares minted in exchange for assets
+    function previewDeposit(uint256 assets) external view returns(uint256) {
+        return _convertToShares(assets);
+    }
+
+    /// @notice used to simulate a deposit at the current block
+    /// @param shares desired amount of shares to mint
+    /// @return assets amount of asset token required for desired shares
+    function previewMint(uint256 shares) external view returns(uint256) {
+        return _convertToAssets(shares);
+    }
+
+    /// @notice used to simulate a withdrawal and the shares required
+    /// @param assets desired amount of asset token to receive
+    /// @return shares amount of shares required to redeem
+    function previewWithdraw(uint256 assets) external view returns(uint256) {
+        return _convertToShares(assets);
+    }
+
+    /**
+     * @dev Initiate a flash loan.
+     * @param receiver The receiver of the tokens in the loan, and the receiver of the callback.
+     * @param token The loan currency.
+     * @param amount The amount of tokens lent.
+     * @param data Arbitrary data structure, intended to contain user-defined parameters.
+     */
+    function flashLoan(
+        IERC3156FlashBorrower receiver,
+        address token,
+        uint256 amount,
+        bytes calldata data
+        ) external override returns (bool) {
+        require(token == address(iRings), "WRings: Unsupported token");
+        uint256 fee_ = _flashFee(amount);
+        require(
+            iRings.transfer(address(receiver), amount),
+            "WRings: Transfer failed"
+        );
+        require(
+            receiver.onFlashLoan(msg.sender, token, amount, fee_, data) == CALLBACK_SUCCESS,
+            "IERC3156: Callback failed"
+        );
+        uint256 _allowance = iRings.allowance(address(receiver), address(this));
+        require(
+            _allowance >= (amount + fee_),
+            "WRings: Repay not approved"
+        );
+        require(
+            iRings.transferFrom(address(receiver), address(this), amount + fee_),
+            "WRings: Repay failed"
+        );
+        return true;
+    }
+
+    function init(uint256 assets, address receiver) external returns(uint256) {
```

Comment:
You probably need a check that `init` can only be called if `totalSupply == 0`. Either that or init in the constructor.